### PR TITLE
Add Discord support

### LINKED ISSUE #97: Add support for Discordapp
**Is your feature request related to a problem? Please describe.**
It would be nice if we can get support for discordapp.com
There is no software like this out there that has support for discord.
There are plenty of API Wrappers for discord out there
(Link to a wrapper for golang:) https://github.com/bwmarrin/discordgo

**Describe the solution you'd like**
Discord has a rich API and Role based secrurity (Eg limiting who can send/read/see channels etc.)

**Additional context**
discordapp.com
https://discordapp.com/developers/applications/

### PR DESCRIPTION
**Pull request by codenio**
#### ISSUE TYPE

 - Feature Pull Request

##### SUMMARY

**Add support for Discord:**

This PR,
- adds discord bot
- adds discord notifier
- adds required discord configurations into comm_config.yaml
- adds discordgo module into go.mod
- adds discord configs into all yaml files

Fixes #97 

### Steps For Creation and Integration:

1. Reach https://discordapp.com/developers/applications 

2. Create New Application named Botkube and add a bot named **Botkube** into the Application

3. Copy the Application **CLIENT ID** and place it under _Dischord.botid_ in comm-config.yaml

4. Copy Bot **TOKEN** and place it under _Dischord.token_ in comm-config.yaml
 
5. Generate the URL with suitable permissions using the OAuth2 URL Generator available under OAuth2 section to the add bot to your Discord server
```
https://discordapp.com/oauth2/authorize?&client_id=<YOUR_CLIENT_ID>&scope=bot&permissions=<SET_OF_PERMISSIONS>
```
6. Paste the generated URL in a new tab, select the discord sever to which you want to add the bot and Authorise Bot addition.
7. Create a new channel and/or copy the  **CHANNEL ID** of an existing one that you want to use and place it under _Dischord.channel_

```yaml
  # Settings for Discord
  discord:
    enabled: true
    token: 'DISCORD_TOKEN'	    # BotKube Bot Token 
    botid: 'DISCORD_BOT_ID'         # BotKube Application Client ID 
    channel: 'DISCORD_CHANNEL_ID'   # Discord Channel id for receiving BotKube alerts 
    notiftype: short                # Change notification type short/long you want to receive. notiftype is optional and Default notification type is short (if not specified)
```

### Preview:

ping
pod-create-short
pod-error-long
file-attachment

### CODE REVIEW

File: cmd/botkube/main.go
Comment (sanketsudake) [inline cmd/botkube/main.go:60]: Can you please rename variable from `db` to `discordBot` ? Db confuses with database.

File: pkg/bot/discord.go
Comment (sanketsudake) [inline pkg/bot/discord.go:20]: Wonder why this is named SlackURL ? Not sure if we need restructring across other providers as well to fix.
cc @PrasadG193

File: pkg/notify/discord.go
Comment (sanketsudake) [inline pkg/notify/discord.go:67]: Missing return error from here ?

File: pkg/notify/discord.go
Comment (sanketsudake) [inline pkg/notify/discord.go:90]: Could we define some constant for time format here ?

File: pkg/notify/discord.go
Comment (codenio) [reply pkg/notify/discord.go:67]: `Send()` in slack,discord and `sendMessage()` in mattermost are implemented in the similar way. do we need to restructure this?

File: pkg/bot/discord.go
Comment (codenio) [reply pkg/bot/discord.go:20]: removed unused variables.

File: pkg/notify/discord.go
Comment (codenio) [reply pkg/notify/discord.go:90]: created a separate const `customTimeFormat` for the same.

File: cmd/botkube/main.go
Comment (codenio) [reply cmd/botkube/main.go:60]: expanded all short forms like sb,mb, and db respectively to maintain uniformity

File: pkg/notify/discord.go
Comment (PrasadG193) [reply pkg/notify/discord.go:70]: Ideally, we should return an error here and log it at caller. We should also include Send method in Bot interface. I've created an issue to track this https://github.com/infracloudio/botkube/issues/270

Comment (codenio) [reply pkg/notify/discord.go:70]: can these changes go with a separate PR?. Can this be considered resolved.?

Comment (PrasadG193) [reply pkg/notify/discord.go:70]: Let's address this in separate PR

Comment (codenio) [reply pkg/notify/discord.go:70]: thanks, @sanketsudake  and @PrasadG193 
marking this as resolved

File: pkg/bot/discord.go
Comment (PrasadG193) [inline pkg/bot/discord.go:1]: Please add License header. You can copy it from other .go files just change the year to 2020

File: pkg/bot/discord.go
Comment (PrasadG193) [inline pkg/bot/discord.go:35]: better to pass config as a parameter to NewDiscordBot() instead of creating new instance each time. Can be handled in https://github.com/infracloudio/botkube/issues/270

File: pkg/bot/discord.go
Comment (PrasadG193) [inline pkg/bot/discord.go:51]: please remove commented code

File: pkg/bot/discord.go
Comment (PrasadG193) [inline pkg/bot/discord.go:49]: Discord, not `slacknot` :) Please fix the comment

File: pkg/bot/discord.go
Comment (PrasadG193) [inline pkg/bot/discord.go:79]: It the comment relevant here? Also please have uniform Bot start log message just like we have for other mediums

File: pkg/bot/discord.go
Comment (PrasadG193) [inline pkg/bot/discord.go:82]: please add comment

File: pkg/notify/discord.go
Comment (PrasadG193) [inline pkg/notify/discord.go:50]:
```suggestion
```

**Comment by PrasadG193** [comment]
@codenio can we add integration tests?

File: pkg/bot/discord.go
Comment (codenio) [reply pkg/bot/discord.go:1]: resolved

File: pkg/bot/discord.go
Comment (codenio) [reply pkg/bot/discord.go:35]: can it be considered for refactoring #270? as this same condition prevails across all the bots

Comment (PrasadG193) [reply pkg/bot/discord.go]: @codenio We have fixed this for other Bots with https://github.com/infracloudio/botkube/pull/272. Could you please address this as a part of this PR only?

Comment (codenio) [reply pkg/bot/discord.go]: @PrasadG193 address those changes and updated the PR yesterday, can you recheck once

**Comment by codenio** [comment]
added discord configs into all yaml files

**Comment by PrasadG193** [comment]
@codenio The CI job is failing with 
```
Step 9/17 : RUN GOOS=${GOOS_VAL} GOARCH=${GOARCH_VAL} go build -o /go/bin/botkube ./cmd/botkube
 ---> Running in 3dcedf3fac47
# github.com/docker/docker/pkg/term
exec: "gcc": executable file not found in $PATH
The command '/bin/sh -c GOOS=${GOOS_VAL} GOARCH=${GOARCH_VAL} go build -o /go/bin/botkube ./cmd/botkube' returned a non-zero code: 2
Makefile:39: recipe for target 'container-image' failed
make: *** [container-image] Error 2
The command "make container-image" exited with 2.
```
I think the docker dep you added need gcc. Can you please update travis config to add gcc dep?

**Comment by codenio** [comment]
@PrasadG193, resolved the build issue and rebased this PR into developed. 
can this be merged?

**Comment by kingdevnl** [comment]
I guess this has to be rebased again soon?

**Comment by codenio** [comment]
rebased and resolved issues

**Comment by unixfox** [comment]
Can this PR get reviewed again or/and merged? I'm using it on my own k8s cluster and it's working pretty well!

@codenio Seriously good job.

**Comment by unixfox** [comment]
I just found an odd bug.
I'm able to ping the bot through the desktop Discord client but not on the Android Discord client.
Capturebotkube
The first message is from the Android Discord client and the second is from the desktop Discord client.

@codenio Could you try to replicate the bug by sending a ping to your botkube from your Android Discord client?

**Comment by kingdevnl** [comment]
How would i install the version of the PR onto a kube cluster?

**Comment by kingdevnl** [comment]
I tried to get it to run,

But gettings this error...

panic: http: multiple registrations for /metrics

goroutine 38 [running]:
net/http.(*ServeMux).Handle(0x3489840, 0x1d1852b, 0x8, 0x216d180, 0xc0002c71d0)
        /usr/local/go/src/net/http/server.go:2391 +0x29c
net/http.Handle(...)
        /usr/local/go/src/net/http/server.go:2434
github.com/infracloudio/botkube/pkg/metrics.ServeMetrics(0x1d0fdda, 0x4)
        /app/pkg/metrics/metrics.go:29 +0x5f
created by main.startController
        /app/cmd/botkube/main.go:81 +0x13d

**Comment by unixfox** [comment]
@kingdevnl I think something is broken in the master branch of botkube.
If you use this commit it will work: https://github.com/infracloudio/botkube/commit/a3ff62f8654cdd2f83c2476f7d96e3f82f09509c

**Comment by kingdevnl** [comment]
@unixfox I did build it using that commit hmmm https://i.imgur.com/DC5duQ4.png

**Comment by PrasadG193** [comment]
@codenio @unixfox we need to come up with a way to add integration tests for this. Otherwise, going forward, as new integration coming in, it would be too difficult to test integrations manually each time we make a release.

**Comment by unixfox** [comment]
@kingdevnl You used the commit 4e0e433 (the newer one) and not a3ff62f (the older one).

**Comment by kingdevnl** [comment]
https://i.imgur.com/OCUC5OX.png

@TestBot get pods
Command not supported. Please run /botkubehelp to see supported commands.

@TestBot ping
pong from cluster 'not-configured'

K8s Server Version: v1.18.3+k3s1
BotKube version: v0.10.0

**Comment by codenio** [comment]
> need to come up with a way to add integration tests for this

@PrasadG193 I guess this is becoming a common issue for notifiers like Mattermost,  Discord, MsTeams, and Telegram.
coming up with a solution would benefit all these notifiers.

creating issues and marking it as Help Wanted would notify more people and might get a helping hand from the community.

**Comment by unixfox** [comment]
@codenio Could you check if you have the same bug as me?

> I just found an odd bug.
> I'm able to ping/send commands to the bot through the desktop Discord client but not on the Android Discord client.
> Capturebotkube
> The first message is from the Android Discord client and the second is from the desktop Discord client.
> 
> @codenio Could you try to replicate the bug by sending a ping to your botkube from your Android Discord client?

**Comment by onedr0p** [comment]
@unixfox @codenio same issue here, only works on desktop client.

Also, is there anyway to send messages as a DM instead of to a channel?

**Comment by codenio** [comment]
@unixfox and @onedr0p, sorry for the delay. 
I had fixed the Android Discord client not working bug and verified that it is working now.
do check and verify the same.

**Comment by unixfox** [comment]
Thank you @codenio!

Does someone know why I'm getting this error when compiling the code with the new commit:
```

INFO[2020-08-07T08:44:27Z] Registering the filter ImageTagChecker       
INFO[2020-08-07T08:44:27Z] Registering the filter IngressValidator      
INFO[2020-08-07T08:44:27Z] Registering the filter NamespaceChecker      
INFO[2020-08-07T08:44:27Z] Registering the filter NodeEventsChecker     
INFO[2020-08-07T08:44:27Z] Registering the filter ObjectAnnotationChecker 
INFO[2020-08-07T08:44:27Z] Registering the filter PodLabelChecker       
INFO[2020-08-07T08:44:27Z] Starting controller                          
INFO[2020-08-07T08:44:27Z] Starting discord bot                         
INFO[2020-08-07T08:44:27Z] BotKube connected to Discord!                
panic: http: multiple registrations for /metrics
goroutine 26 [running]:
net/http.(*ServeMux).Handle(0x3489840, 0x1d1852b, 0x8, 0x216d1e0, 0xc0001cf560)
    /usr/local/go/src/net/http/server.go:2391 +0x29c
net/http.Handle(...)
    /usr/local/go/src/net/http/server.go:2434
github.com/infracloudio/botkube/pkg/metrics.ServeMetrics(0xc00004004d, 0x4)
    /app/pkg/metrics/metrics.go:29 +0x5f
created by main.startController
    /app/cmd/botkube/main.go:81 +0x13d
stream closed
```
?

**Comment by codenio** [comment]
I got this error, but managed to bypass it by commenting few lines in main.go (line 81 and above) for fixing the bug

/metrics endpiont used by prometheus client for exposing metrics is getting multiple http registrations.

**Comment by onedr0p** [comment]
@PrasadG193 are unit tests really holding up adding new notifier integrations, specifically this one? I understand tests are important, however they can always be added later on.

**Comment by codenio** [comment]
resolved it. can you check now..?

**Comment by unixfox** [comment]
Yes it's working without any error. Thank you!

**Comment by onedr0p** [comment]
Working great on mobile now too with @unixfox docker image. Thanks you two!

**Comment by onedr0p** [comment]
Any update here from the maintainers? I've been using a docker image built from this PR for a month now and it works great.

**Comment by PrasadG193** [comment]
@codenio We can target and get this in in upcoming release. Could you please take a stab at adding documentation? We use hugo for hosting docs. The docs repo is: https://github.com/infracloudio/botkube-docs

**Comment by PrasadG193** [comment]
We have improved test coverage for the notifier interface. For discord bot interface tests, we can revisit this later.

**Comment by codenio** [comment]
@PrasadG193 will certainly add corresponding docs, can we get this PR merged now..?

File: comm_config.yaml
Comment (PrasadG193) [inline comm_config.yaml:15]: Please fix the comment indentation. Let's have a uniform indentation for all the comments

File: comm_config.yaml
Comment (PrasadG193) [inline comm_config.yaml:30]:
```suggestion
    token: 'DISCORD_TOKEN'	       # BotKube Bot Token 
```

Comment (PrasadG193) [reply comm_config.yaml:30]: Fix indentation

File: deploy-all-in-one-tls.yaml
Comment (PrasadG193) [inline deploy-all-in-one-tls.yaml:263]: Fix indentation here and similar other occurances

File: pkg/bot/discord.go
Comment (PrasadG193) [inline pkg/bot/discord.go:28]:
```suggestion
	"strings"

	"github.com/bwmarrin/discordgo"
	
	"github.com/infracloudio/botkube/pkg/config"
	"github.com/infracloudio/botkube/pkg/execute"
	"github.com/infracloudio/botkube/pkg/log"
```
Let's divide the import packages into 3 groups - 
```
built-in

3rd-party

botkube packages
```

File: pkg/bot/discord.go
Comment (PrasadG193) [inline pkg/bot/discord.go:147]: We should log request message also. I know this is missing in other Bot interface implementations, but this will make things debug easier

File: pkg/bot/discord.go
Comment (codenio) [reply pkg/bot/discord.go:28]: done

File: deploy-all-in-one-tls.yaml
Comment (codenio) [reply deploy-all-in-one-tls.yaml:263]: done

File: comm_config.yaml
Comment (codenio) [reply comm_config.yaml:30]: done

File: comm_config.yaml
Comment (codenio) [reply comm_config.yaml:15]: done

File: pkg/bot/discord.go
Comment (codenio) [reply pkg/bot/discord.go:147]: added a debug log for the same after INFO `log.Debugf("Dumped responce : %+v", dm.response)`

File: deploy-all-in-one-tls.yaml
Comment (PrasadG193) [inline deploy-all-in-one-tls.yaml:264]: Please fix the indentation.

File: deploy-all-in-one.yaml
Comment (PrasadG193) [inline deploy-all-in-one.yaml:264]: same here..

File: helm/botkube/values.yaml
Comment (PrasadG193) [inline helm/botkube/values.yaml:300]: here too

File: test/comm_config.yaml
Comment (PrasadG193) [inline test/comm_config.yaml:36]: Please fix the indentation. I think you are using mix of tabs and spaces. Let's stick with spaces only

Comment (codenio) [reply test/comm_config.yaml]: my editor was dint show me the difference, however fixed these some how

File: pkg/bot/discord.go
Comment (codenio) [reply pkg/bot/discord.go:149]: added Info level to log request sent to botkube, and the corresponding valid responses generated by botkube.

Comment (codenio) [reply pkg/bot/discord.go:149]: created a separate PR #404, for adding valid Request and Response log, and to remove capture of all events in slack.

**Review by PrasadG193 — APPROVED** [review]

**Comment by codenio** [comment]
:pray: , feeling relieved., @all thanks for the extended support, thanks @PrasadG193 and @sanketsudake for the reviews and feedbacks.

**Comment by PrasadG193** [comment]
Thanks for all the hard work and patience @codenio

**Comment by kingdevnl** [comment]
Realy thanks man  <3

**Pull request merged**
